### ISSUE TITLE
Issues with Polar Conversion

### ISSUE THREAD

**Issue by gmullac** [dpenney/grbl-1.1-polr#2]
I am a newbie so I hope someone is willing to answer what I assume is going to be a fairly obvious issue - though I have spent several late nights trying to figure it out. 

I have loaded this firmware grbl-1.1-polar onto my Arduino Uno with a gShield. I have two NEMA 17 motors mounted and wired. I am using the latest version of UGS to send the gcode commands. 

The issue I seem to be having is that the commands are not being modified from cartesian coordinates to polar coordinates. My understanding is that when I send a jog command of Y100 that both the X and Y stepper should become active. Currently when  Y100 code is sent only the Y stepper moves. It is almost as if standard cartesian grbl is running. 

I have tried everything I can think of - reinstalled arduino, deleted all arduino libraries and reinstalled only grbl-1.1-polr. I haven't changed anything in config.h or any other files. Are there other compile changes that need to occur?

Any ideas would be appreciated. I am not sure where to go next.

**Comment by dpenney** [comment]
If this is for a drawbot send it to the mid point on your x axis first. So if your motors are 1000mm apart send to x500 y500. From there try and move to x500 y300 and both motors should do the same thing i.e making both cords shorter.

Also G0 will end at the right spot but won't travel in a straight line, G1 will use polar to track straight lines.

Hope that helps.

**Comment by dpenney** [comment]
Also after sending to x500 y500, manually adjust the pen so it's actually centered and 500mm down from the motors.

**Comment by gmullac** [comment]
Thanks so much for responding.

In my build, the motors are 570mm apart so I moved to X285 Y285. I then made sure the pen holder was at this location on the board. After commanding G1 X285 Y200, only the Y stepper moved approx. 85mm. I then repeated the setup to move the pen holder to X285 Y285 again and commanded G1 X200 Y285 and the X stepper motor became active. I confirmed that the defaults file was updated so the distance was 570. 

It appears for some reason that the commands are being sent as cartesian commands and not being converted to string length changes.

**Comment by dpenney** [comment]
Definitely sounds like you aren't in polar mode - and I take it the upload went okay and POLAR was defined?

[motion_control.c](https://github.com/dpenney/grbl-1.1-polr/commit/aaf89b61199dde53bdc5178afc6f342c96389d21#diff-0313c87fa5fb427fd27970cfd83f39bf70e46eb16e2330062af840f771d1f2f5) is where the actual transformation occurs, as long as it's compiled with POLAR defined this should come into play, it takes the target x and y, converts to polar co-ords, and then passes it off to the planner.

When you view the grbl settings via $$ does everything look okay - specifically the distance?

**Comment by gmullac** [comment]
I think this is leading down the right path. It doesn't appear that any of the default values correlate to the values when I check the grbl settings via $$

When I tried to display grbl settings the values displayed do not even include the distance. Distance should fall under $28 (I think) but that isn't displayed. Also when I change other values, such as steps/mm those values don't reflect my defaults either. It seems the defaults are being set by some other location.  

![image](https://user-images.githubusercontent.com/39268828/102568172-fa23bd00-40b0-11eb-9f71-20dd52490ec9.png)

I copied a screenshot of the $$ output. The values should be what is in the downloaded defaults. 

I did try to compile without the default.h file and it didn't compile so it appears to be looking in the correct location. 
I looked through motion_control.c and it appears to have the defines correct and config.h also looks correct.

Thanks again for your help.

**Comment by dunkfordyce** [comment]
I have also tried this fork and Im sure there is a bunch of code missing or uncommitted because comparing it to the original large parts are missing or wrong.

**Comment by dpenney** [comment]
I don't think anything is missing as this is the version that I used on my plotter.

Perhaps try this https://github.com/ilaro-org/grbl-polar - it uses an older version of grbl but it's where the code to do the polar conversion came from.

**Comment by dpenney** [comment]
Also if you add some debug statements before and after the polar conversion do you see this version changing the x/y locations?

I'll recompile this and double check it.

**Comment by dunkfordyce** [comment]
target_polar isnt even used here unless im mistaken 
https://github.com/dpenney/grbl-1.1-polr/blob/master/grbl/motion_control.c#L71

**Comment by dpenney** [comment]
Hmmm.... you might be right...  try changing this line in motion_control.c

if (plan_buffer_line(target, pl_data) == PLAN_EMPTY_BLOCK) {
to
if (plan_buffer_line(target_polar, pl_data) == PLAN_EMPTY_BLOCK) {

I'm still setting this project back up to test.

**Comment by dpenney** [comment]
So I'd actually moved to an ESP32 based project - update to this and you should be okay, looks like it didn't get back ported to this arduino version.

#ifdef POLAR

    //Change from cartessian to polar coordinates
    float target_polar[N_AXIS];
    float x= settings.distance-target[X_AXIS];
    target_polar[X_AXIS]=sqrt(labs(target[X_AXIS]*target[X_AXIS]+target[Y_AXIS]*target[Y_AXIS]));
    target_polar[Y_AXIS]=sqrt(labs(x*x+target[Y_AXIS]*target[Y_AXIS]));
    target_polar[Z_AXIS]=0.0;
    plan_buffer_line(target_polar, feed_rate, invert_feed_rate);

    gc_state.position[X_AXIS]=target[X_AXIS];
    gc_state.position[Y_AXIS]=target[Y_AXIS];

#else
        plan_buffer_line(target_polar, feed_rate, invert_feed_rate);
#endif

**Comment by dunkfordyce** [comment]
Which esp32 project? polargraph_server_polarshield_esp32?

**Comment by dpenney** [comment]
Okay so I just pushed a commit - please try again.

ESp32 version was based off of this https://github.com/bdring/Grbl_Esp32 with the same polar updates that are used here
https://github.com/bdring/Grbl_Esp32 - I never published it but it's what I ended up using along with this shield https://www.hackster.io/news/you-can-now-run-grbl-on-your-esp32-ba689909fa0c - very nice being able to send the gcode over bluetooth.

**Comment by dunkfordyce** [comment]
Well that seems to make two steppers turn. Ill try hanging it up...

**Comment by gmullac** [comment]
I just tried it myself as well and both steppers turn, however, commanding a X100 move and commanding a Y100 move result in nearly the same motion.

**Comment by dunkfordyce** [comment]
Yea it doesnt work.

**Comment by dpenney** [comment]
If you are at your center location on the X axis you would expect both steppers to turn the same amount in opposite directions if you tell it to move along the Y axis, i.e it will shorten/lengthen the cords on both sides which will raise the pointer in a straight line.

This assumes you have $27 I think it is, set to the distance between motors.
$27=width in mm between steppers
$X to unlock
So first send to G0 x width/2 y width/2
Move your pen manually to the center position once steppers stop

G1 F1000
G1 x width/2 Y300 will result in both steppers moving the same amount in opposite directions pulling the pen up in a straight line. You might have to tweak the stepper mask if the steppers move in the same direction.

**Comment by dpenney** [comment]
Again feel free to install the grbl 0.9 project I linked to and test using that, all the polar code comes from there.

**Comment by dunkfordyce** [comment]
I can run the 0.9 code on my arduino but I cant connect to it with
Universal GCode sender. Maybe I need an older version

**Comment by dpenney** [comment]
If you are seeing both motors moving then you're on the right track as that is what you would expect.

Here are the setup instructions written for the science fair project this was built for, perhaps they will be of assistance.

Setup
Once the Drawbot is assembled you will need to make sure the settings are correct so it can draw nice straight lines.

Upon startup you should tell the machine to go home. For a Polar machine the X axis home position is the spot in the middle of the two servos. So if your servos are 1000mm apart, X home will be 500m. Y home is a bit more flexible but the best spot but using the same distance as X usually works pretty well.

So find that home spot on your machine and enter 
G0 X500 Y500 
After the servos stop moving adjust the belts so that the pen is on the home spot.

The next thing to test is the Steps per mm setting.
If you enter G0 x550 your pen should move 50mm, and hopefully it moved to the right. If it doesn’t you need to change the steps per mm setting which is
$100 - X Servo 
$101 - Y Servo

If the distance your pen moved is too small you need to increase the steps per mm.
$100 = 75 and $101 = 75 work on our machine.

Bring the machine back to home

Now we are going to make sure our stepper motors are moving in the right direction.
If you enter G0 x650 you should the the left belt get longer and the right belt get shorter. If it didn’t you can flip the direction using the $3 command

$3 = 0 - is the default
$3 = 1 will reverse the direction of the X servo
$3 = 2 will reverse the direction of the Y servo
$3 = 3 will reverse the directions of both servos

Use whichever setting makes the left belt get longer and the right belt get shorter when you go from G0 X500 Y500 to G0 X600 Y500.

**Comment by dpenney** [comment]
> c

You will also need to make sure that you have the correct settings for $100 and $101 - how many steps to move 1mm.

**Comment by gmullac** [comment]
I finally had some time to sit down and look at the changes. I completed the steps of uploading the code directly from the source but it still didn't work. The result was much better though - the motors were working in unison but the issue was that X and Y commands completed the same output from the motors. I looked through the code and couldn't find any obvious issues. I tried hardcoding the distance setting instead of looking for the distance in the defaults and surprise! the machine started working as expected. Apparently, there is an issue referencing the defaults, I am not sure if this is specific to my instance or if it is an issue with the code in general. 

I still haven't dialed in all the settings but I don't think there will be any more issues. All in all, I consider this issue closed. Thanks for your help.
